File: java/src/test/java/lexer/TokenTest.java
package lexer;

import common.PeekIterator;
import org.junit.Test;

import java.util.HashMap;
import java.util.Map;

import static org.junit.Assert.*;

public class TokenTest {

    @Test
    public void test_makeVarOrKeyWord() {
        PeekIterator<Character> it1 = new PeekIterator<>("if abc".chars().mapToObj(x -> (char) x));
        PeekIterator<Character> it2 = new PeekIterator<>("true abc".chars().mapToObj(x -> (char) x));

        Token token1 = Token.makeVarOrKeyWord(it1);
        Token token2 = Token.makeVarOrKeyWord(it2);

        assertToken(token1, TokenType.KEYWORD, "if");
        assertToken(token2, TokenType.BOOLEAN, "true");

        it1.next();

        Token token3 = Token.makeVarOrKeyWord(it1);
        assertToken(token3, TokenType.VARIABLE, "abc");

    }

    private void assertToken(Token token1, TokenType keyword, String s) {
        assertEquals(keyword, token1.getType());
        String value = token1.getValue();
        assertEquals(s, value);
    }

    @Test
    public void test_makeString() {
        String[] tests = {
                "\"abc\"",
                "\'abc\'",
                "\"\"",
                "\'\'",
        };
        try {

            for (String test : tests) {
                PeekIterator<Character> it1 = new PeekIterator<>(test.chars().mapToObj(c -> (char) c));
                Token token = Token.makeString(it1);
                assertToken(token, TokenType.STRING,test);
            }

        } catch (LexicalException e) {
            e.printStackTrace();
        }
    }

    @Test
    public void test_makeOp() {
        String[] tests = {
                "+ xxx",
                "++mmm",
                "/=g",
                "==1",
                "&=3982",
                "&777",
                "||xxx",
                "^=111",
                "%7"
        };
        String[] results = {"+","++","/=","==","&=","&","||","^=","%"};

        try {

            for (int i = 0; i < tests.length; i++) {
                PeekIterator<Character> it = new PeekIterator<>(tests[i].chars().mapToObj(c->(char)c));
                Token token = Token.makeOp(it);
                assertToken(token, TokenType.OPERATOR,results[i]);
            }

        } catch (LexicalException e) {
            e.printStackTrace();
        }

    }

    @Test
    public void test_makeNumber() {
        String[] tests={
                "2",
                "-2",
                "+0 aa",
                "-0 aa",
                ".3 ccc",
                ".555 ddd",
                "7789.8888 ddd",
                "-1000.123123*123123",
        };
        try {
            for (String test : tests) {
                PeekIterator<Character> it = new PeekIterator<>(test.chars().mapToObj(c -> (char) c),(char)0);
                Token token = Token.makeNumber(it);

                String[] split = test.split("[* ]+");
                String result = split[0];

                assertToken(token, result.contains(".") ? TokenType.FLOAT : TokenType.INTEGER, result);
            }
        } catch (LexicalException e) {
            System.out.println(e.getMessage());
            e.printStackTrace();
        }
    }
}

File: js/test/Token.test.js
const {assert}  =require("chai")

const arrayToGenerator  =require("../common/arrayToGenerator")
const PeekIterator  =require("../common/PeekIterator")
const TokenType  =require("../lexer/TokenType")
const Token = require("../lexer/Token");

describe('Token', ()=> {

    function assertToken(token,type,value){
        assert.equal(token.getType(),type)
        assert.equal(token.getValue(),value)
    }

    it('varOrKeyWord',  ()=> {

        let it1 = new PeekIterator(arrayToGenerator([..."if abc"]))
        let it2 = new PeekIterator(arrayToGenerator([..."true abc"]))

        let token1 = Token.makeVarOrKeyWord(it1);
        let token2 = Token.makeVarOrKeyWord(it2);

        it1.next()
        let token3 = Token.makeVarOrKeyWord(it1);

        assertToken(token1,TokenType.KEYWORD,"if")
        assertToken(token2,TokenType.BOOLEAN,"true")
        assertToken(token3,TokenType.VARIABLE,"abc")
    });

    it('makeString', function () {
        let arr=[
            `"123"`,
            `'123'`,
            `'abc'`,
            `''`,
            `""`,
        ]
        let res=[
            `"123"`,
            `'123'`,
            `'abc'`,
            `''`,
            `""`,
        ]

        for (let i = 0; i < arr.length; i++) {
            let it = new PeekIterator(arrayToGenerator([...arr[i]]));
            let token = Token.makeString(it);
            assertToken(token,TokenType.STRING,res[i])
        }
    });

    it('makeOp',  () => {

        const tests=[
            ["+ xxx","+"],
            ["++mmm","++"],
            ["/=g","/="],
            ["==1","=="],
            ["&=3982","&="],
            ["&777","&"],
            ["||xxx","||"],
            ["^=111","^="],
            ["%7","%"],
        ]

        for (let test of tests) {
            const [input,expected] =test
            let it = new PeekIterator(arrayToGenerator([...input]));
            let token = Token.makeOp(it);
            assertToken(token,TokenType.OPERATOR,expected);
        }
    });

    it('makeNumber', ()=> {
        let tests=[
            ['0.12asd','0.12'],
            ['+0.12asd','+0.12'],
            ['-0.12asd','-0.12'],
            ['+12asd','+12'],
            ['.12asd','.12'],
        ];

        for (const test of tests) {
            const [input,expected]=test
            let it = new PeekIterator(arrayToGenerator([...input]));
            let token = Token.makeNumber(it);

            let tokenType = token.getValue().indexOf(".")===-1?TokenType.INTEGER: TokenType.FLOAT;
            assertToken(token,tokenType,expected)
        }
    });
})